File: templates/problems/15_greedy/04_gas_station/gas_station_test.go
package gas_station

import "testing"

func TestCanCompleteCircuit(t *testing.T) {
	tests := []struct {
		name     string
		gas      []int
		cost     []int
		expected int
	}{
		{"example 1", []int{1, 2, 3, 4, 5}, []int{3, 4, 5, 1, 2}, 3},
		{"no solution", []int{2, 3, 4}, []int{3, 4, 3}, -1},
		{"single station enough", []int{5}, []int{3}, 0},
		{"single station not enough", []int{2}, []int{4}, -1},
		{"start at index 0", []int{3, 1, 1}, []int{1, 2, 2}, 0},
		{"all equal", []int{1, 1, 1}, []int{1, 1, 1}, 0},
		{"start at last", []int{1, 1, 5}, []int{2, 3, 1}, 2},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := canCompleteCircuit(tt.gas, tt.cost)
			if got != tt.expected {
				t.Errorf("canCompleteCircuit(%v, %v) = %d, want %d", tt.gas, tt.cost, got, tt.expected)
			}
		})
	}
}

File: templates/problems/15_greedy/04_gas_station/typescript/gasStation.test.ts
import { describe, it, expect } from 'vitest';
import { canCompleteCircuit } from './gasStation';

describe('GasStation', () => {
    it('example 1', () => {
        expect(canCompleteCircuit([1, 2, 3, 4, 5], [3, 4, 5, 1, 2])).toBe(3);
    });

    it('no solution', () => {
        expect(canCompleteCircuit([2, 3, 4], [3, 4, 3])).toBe(-1);
    });

    it('single station enough', () => {
        expect(canCompleteCircuit([5], [3])).toBe(0);
    });

    it('single station not enough', () => {
        expect(canCompleteCircuit([2], [4])).toBe(-1);
    });

    it('start at index 0', () => {
        expect(canCompleteCircuit([3, 1, 1], [1, 2, 2])).toBe(0);
    });

    it('all equal', () => {
        expect(canCompleteCircuit([1, 1, 1], [1, 1, 1])).toBe(0);
    });

    it('start at last', () => {
        expect(canCompleteCircuit([1, 1, 5], [2, 3, 1])).toBe(2);
    });
});
